cluster/v1beta1/types_managedclusterset.go

```diff
@@ -38,6 +38,41 @@ type ManagedClusterSet struct {
 
 // ManagedClusterSetSpec describes the attributes of the ManagedClusterSet
 type ManagedClusterSetSpec struct {
+	// ClusterSelector represents a selector of ManagedClusters
+	// +optional
+	ClusterSelector ManagedClusterSelector `json:"clusterSelector,omitempty"`
+}
+
+// ManagedClusterSelector represents a selector of ManagedClusters
+type ManagedClusterSelector struct {
+	// SelectorType could only be ExclusiveLabel now, will support more SelectorType later
+	// "ExclusiveLabel" means to use a particular cluster label. It guarantees that clustersets with same label key are exclusive
+	// +kubebuilder:validation:Enum=ExclusiveLabel
+	// +kubebuilder:default:="ExclusiveLabel"
+	// +optional
+	SelectorType SelectorType `json:"selectorType"`
+
+	// ExclusiveLabel defines one label which clusterset could use to select target managedClusters. In this way, we will:
+	// 1. Guarantee clustersets with same label key are exclusive
+	// 2. Enable additional permission check when cluster joining/leaving a clusterset.
+	// Currentlly, When a user want to add a label "cluster.open-cluster-management.io/clusterset:<ManagedClusterSet Name>" for managedCluster,
+	// he/she should have "create" permission for subresource "managedclusters/label" with name "cluster.open-cluster-management.io/clusterset:<ManagedClusterSet Name>".
+	// +optional
+	ExclusiveLabel *ManagedClusterLabel `json:"exclusiveLabel,omitempty"`
+}
+
+type SelectorType string
+
+const (
+	ExclusiveLabel SelectorType = "ExclusiveLabel"
+)
+
+//ManagedClusterLabel defines one label of managedCluster
+type ManagedClusterLabel struct {
```

Comment:
Thinking about the API, I like this way of describing it, **but** this would require a new API version and a conversionwebhook set it to the .metadata.name value.  That's a lot of overhead.

If we were willing to limit the configuration available, we could instead have a `LegacyClusterSetLabel` value in the discriminated union with no additional configuration.  The `LegacyClusterSetLabel` would have the matching behave the way it functions today.  This would allow us to use the built-in defaulting inside the same API version without any additional webhook logic.